cmd/system-shutdown/system-shutdown.c

```diff
@@ -93,7 +84,33 @@ int main(int argc, char *argv[])
 		}
 	}
 
-	reboot(cmd);
+	// We are reading a file from /run and need to do this before unmounting
+	if (cmd == RB_AUTOBOOT)
+		read_reboot_arg(reboot_arg, sizeof reboot_arg);
+
+	if (umount_all()) {
+		kmsg("- found no hard-to-unmount writable partition.");
+	} else {
+		if (mount("/oldroot/writable", "/writable", NULL, MS_MOVE, NULL)
+		    < 0) {
+			die("cannot move writable out of the way");
+		}
+
+		bool ok = umount_all();
+		kmsg("%c was %s to unmount writable cleanly", ok ? '-' : '*',
+		     ok ? "able" : "*NOT* able");
+		sync();		// shouldn't be needed, but just in case
+	}
+
+	// glibc reboot wrapper does not expose the optional reboot syscall
+	// parameter
+
+	if (cmd == RB_AUTOBOOT && reboot_arg[0] != '\0')
+		syscall(SYS_reboot,
+			LINUX_REBOOT_MAGIC1, LINUX_REBOOT_MAGIC2,
+			LINUX_REBOOT_CMD_RESTART2, reboot_arg);
+	else
+		reboot(cmd);
```

Comment:
Can you add error handling to both cases please. It doesn't have to be very sophisticated but I'd like to see it.